File: services/22-rails/spec/requests/auth_spec.rb
require 'rails_helper'

RSpec.describe AuthController, type: :request do
  let(:user) do
    User.create!(
      email: 'test@example.com',
      name: 'Test User',
      provider: 'github'
    )
  end

  let(:valid_token) do
    JwtService.encode({ sub: user.id, email: user.email, name: user.name })
  end

  describe 'GET /auth/me' do
    context 'with no auth header' do
      it 'returns 401' do
        get '/auth/me'
        expect(response).to have_http_status(:unauthorized)
        expect(JSON.parse(response.body)['error']).to eq('Unauthorized')
      end
    end

    context 'with a valid JWT Bearer token' do
      it 'returns 200 and current user' do
        get '/auth/me', headers: { 'Authorization' => "Bearer #{valid_token}" }
        expect(response).to have_http_status(:ok)
        body = JSON.parse(response.body)
        expect(body.dig('user', 'email')).to eq('test@example.com')
      end
    end

    context 'with a valid X-API-Key header' do
      it 'returns 200 and current user' do
        get '/auth/me', headers: { 'X-API-Key' => user.api_key }
        expect(response).to have_http_status(:ok)
        body = JSON.parse(response.body)
        expect(body.dig('user', 'email')).to eq('test@example.com')
      end
    end

    context 'with an invalid JWT' do
      it 'returns 401' do
        get '/auth/me', headers: { 'Authorization' => 'Bearer not.a.real.token' }
        expect(response).to have_http_status(:unauthorized)
      end
    end

    context 'with a wrong API key' do
      it 'returns 401' do
        get '/auth/me', headers: { 'X-API-Key' => 'wrong-key' }
        expect(response).to have_http_status(:unauthorized)
      end
    end
  end

  describe 'POST /auth/logout' do
    it 'returns 200 regardless of auth state' do
      post '/auth/logout', headers: { 'Authorization' => "Bearer #{valid_token}" }
      expect(response).to have_http_status(:ok)
    end
  end

  describe 'POST /auth/api-key' do
    context 'with valid auth' do
      it 'returns a new api_key' do
        post '/auth/api-key', headers: { 'Authorization' => "Bearer #{valid_token}" }
        expect(response).to have_http_status(:ok)
        body = JSON.parse(response.body)
        expect(body['api_key']).to be_present
        expect(body['api_key']).not_to eq(user.api_key)
      end
    end

    context 'without auth' do
      it 'returns 401' do
        post '/auth/api-key'
        expect(response).to have_http_status(:unauthorized)
      end
    end
  end

  describe 'DELETE /auth/api-key' do
    context 'with valid auth' do
      it 'revokes the api_key' do
        delete '/auth/api-key', headers: { 'Authorization' => "Bearer #{valid_token}" }
        expect(response).to have_http_status(:ok)
        user.reload
        expect(user.api_key).to be_nil
      end
    end

    context 'without auth' do
      it 'returns 401' do
        delete '/auth/api-key'
        expect(response).to have_http_status(:unauthorized)
      end
    end
  end
end

File: services/api-nestjs/tests/auth.spec.ts
// Auth tests for NestJS.
// Tests the AuthGuard using NestJS testing module.
// Prisma is mocked via jest.mock.

import { Test, TestingModule } from '@nestjs/testing'
import { INestApplication } from '@nestjs/common'
import * as request from 'supertest'
import jwt from 'jsonwebtoken'
import { AppModule } from '../src/app.module'

// ── Mock Prisma ────────────────────────────────────────────────────────────
const mockUser = { id: 'user-1', email: 'test@example.com', name: 'Test User', apiKey: 'yar_testkey123' }

jest.mock('@prisma/client', () => ({
  PrismaClient: jest.fn().mockImplementation(() => ({
    user: {
      findUnique: jest.fn().mockImplementation(({ where }: { where: { apiKey?: string; id?: string } }) => {
        if (where.apiKey === 'yar_testkey123') return Promise.resolve(mockUser)
        if (where.id === 'user-1') return Promise.resolve(mockUser)
        return Promise.resolve(null)
      }),
      upsert: jest.fn().mockResolvedValue(mockUser),
      update: jest.fn().mockResolvedValue(mockUser),
    },
    item: {
      findMany: jest.fn().mockResolvedValue([]),
      create: jest.fn().mockResolvedValue({ id: 'item-1', title: 'Test', userId: 'user-1' }),
      findFirst: jest.fn().mockResolvedValue(null),
      update: jest.fn().mockResolvedValue({}),
      delete: jest.fn().mockResolvedValue({}),
    },
  })),
}))

const JWT_SECRET = process.env.JWT_SECRET ?? 'dev-secret-change-in-production'
const validToken = jwt.sign({ id: mockUser.id, email: mockUser.email, name: mockUser.name }, JWT_SECRET, { expiresIn: '1h' })

describe('GET /auth/me — AuthGuard', () => {
  let app: INestApplication

  beforeAll(async () => {
    const mod: TestingModule = await Test.createTestingModule({ imports: [AppModule] }).compile()
    app = mod.createNestApplication()
    await app.init()
  })

  afterAll(() => app.close())

  it('returns 401 when no auth provided', async () => {
    const res = await request(app.getHttpServer()).get('/auth/me')
    expect(res.status).toBe(401)
  })

  it('returns 401 when JWT is invalid', async () => {
    const res = await request(app.getHttpServer()).get('/auth/me').set('Authorization', 'Bearer bad.token')
    expect(res.status).toBe(401)
  })

  it('returns 200 with valid JWT', async () => {
    const res = await request(app.getHttpServer()).get('/auth/me').set('Authorization', `Bearer ${validToken}`)
    expect(res.status).toBe(200)
    expect(res.body.user.email).toBe(mockUser.email)
  })

  it('returns 200 with valid API key', async () => {
    const res = await request(app.getHttpServer()).get('/auth/me').set('X-API-Key', 'yar_testkey123')
    expect(res.status).toBe(200)
    expect(res.body.user.email).toBe(mockUser.email)
  })

  it('returns 401 with invalid API key', async () => {
    const res = await request(app.getHttpServer()).get('/auth/me').set('X-API-Key', 'yar_invalid')
    expect(res.status).toBe(401)
  })
})
